staging/src/k8s.io/component-helpers/node/util/ips.go

```diff
@@ -52,20 +51,22 @@ func parseNodeIP(nodeIP string, allowDual, sloppy bool) ([]net.IP, error) {
 	}
 
 	if len(nodeIPs) > 2 || (len(nodeIPs) == 2 && netutils.IsIPv6(nodeIPs[0]) == netutils.IsIPv6(nodeIPs[1])) {
-		return nil, fmt.Errorf("must contain either a single IP or a dual-stack pair of IPs")
+		return nil, invalidIPs, fmt.Errorf("must contain either a single IP or a dual-stack pair of IPs")
 	} else if len(nodeIPs) == 2 && !allowDual {
-		return nil, fmt.Errorf("dual-stack not supported in this configuration")
+		return nil, invalidIPs, fmt.Errorf("dual-stack not supported in this configuration")
 	} else if len(nodeIPs) == 2 && (nodeIPs[0].IsUnspecified() || nodeIPs[1].IsUnspecified()) {
-		return nil, fmt.Errorf("dual-stack node IP cannot include '0.0.0.0' or '::'")
+		return nil, invalidIPs, fmt.Errorf("dual-stack node IP cannot include '0.0.0.0' or '::'")
 	}
 
-	return nodeIPs, nil
+	return nodeIPs, invalidIPs, nil
 }
 
-// ParseNodeIPArgument parses kubelet's --node-ip argument. If nodeIP contains invalid
-// values, they will be logged and ignored. Dual-stack node IPs are allowed if
-// cloudProvider is unset or `"external"`.
-func ParseNodeIPArgument(nodeIP, cloudProvider string) ([]net.IP, error) {
+// ParseNodeIPArgument parses kubelet's --node-ip argument.
+// If nodeIP contains invalid values, they will be returned as strings.
```

Comment:
this API now is really weird, I assume that is to avoid logging inside a library that ... IIRC we have to do because of the sloppy IPs thing, and is only used in this file and in cmd/kubelet so I think that is not as bad ... @danwinship WDYT?

Reply:
I added `parseNodeIP` originally but I feel like sig-node owns it now so it's more up to them.

The logging-in-the-middle-of-a-helper-function was just to preserve the original behavior where it used to be logging directly from `cmd/kubelet/app/server.go`, and probably I wouldn't have done it that way if I'd had to think about contexts/loggers at the time.

One possibility would be to not change the API of `parseNodeIPs`, and just have `RunKubelet` check if `len(nodeIPs) == len(strings.Split(kubeServer.NodeIP, ","))` and warn that "some values" were invalid and were ignored if not.

Reply:
Returning `warning` (presumably a string) and a non-fatal pre-formatted error in the second case both have the problem that they use unstructured formatting (i.e. some flavor of fmt.Sprintf) of a message. That might not work well when the caller uses structured logging. I prefer returning just the information about a problem and leaving the formatting of that to the caller.

`len(nodeIPs) == len(strings.Split(kubeServer.NodeIP, ","))` feels like it defeats the purpose of abstracting the parsing in a helper function.

> I added parseNodeIP originally but I feel like sig-node owns it now so it's more up to them.

They'll probably just defer to you :stuck_out_tongue_winking_eye: 

Do want to pull someone else in or do you feel confident going ahead with it as-is and approving it?

Reply:
> Maybe we need a standard convention for functions that want to emit/return both warnings and errors?

Defining conventions is hard, ensuring that fellow developers know about them and follow them is even harder. I suspect unless Go itself defines something, we won't have a chance to establish this on our own for Kubernetes.

Reply:
@mrunalp: you reviewed this earlier for SIG Node. Any opinion ("I feel like sig-node owns it now so it's more up to them")? ^^^

Reply:
> Do you want to pull someone else in or do you feel confident going ahead with it as-is and approving it?

I don't have approver bits on those directories anyway...

> I prefer returning just the information about a problem and leaving the formatting of that to the caller.

I guess that since the semantics of the function are essentially "parse the valid IPs but ignore the invalid ones for backward-compatibility reasons", then returning a list of the invalid IPs does make sense.

Reply:
> I don't have approver bits on those directories anyway...

@mrunalp already approved earlier. What I need from you (or someone from SIG Networking) is approval for [pkg/controller/nodeipam/ipam/OWNERS](https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/nodeipam/ipam/OWNERS) - I know, it's complicated :cry: